docs/naaman.8
define(MONTH_YEAR,)
define(VERSION,)
.TH NAAMAN "1" "MONTH_YEAR" "naaman (VERSION)" "User Commands"
.SH NAME
naaman \- Not Another Aur MANager
.SH DESCRIPTION
usage: naaman [\-h] [\-S] [\-R] [\-Q] [\-u] [\-s] [\-c] [\-d] [\-\-version] [\-\-no\-sudo]
.IP
[\-\-verbose] [\-\-trace] [\-\-pacman PACMAN] [\-\-config CONFIG]
[\-\-no\-confirm] [\-q] [\-\-cache\-dir CACHE_DIR] [\-\-no\-config]
[\-\-builds BUILDS] [\-\-ignore\-for N [N ...]]
[\-\-vcs\-ignore VCS_IGNORE] [\-\-no\-vcs] [\-y] [\-yy]
[\-\-ignore N [N ...]] [\-\-no\-cache] [\-\-skip\-deps] [\-\-reorder\-deps]
[\-\-rpc\-cache RPC_CACHE] [\-i] [\-\-vcs\-install\-only] [\-yyy] [\-f]
[\-\-fetch\-dir FETCH_DIR]
[\-\-rpc\-field {name\-desc,name,maintainer}]
[\-\-do\-not\-track N [N ...]] [\-\-makedeps] [\-g]
.SS "optional arguments:"
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-S\fR, \fB\-\-sync\fR
synchronize packages (install/update). a sync
operation will attempt to install and/or update
packages. \fB\-S\fR by itself will attempt to install a list
of target packages.
.TP
\fB\-R\fR, \fB\-\-remove\fR
remove a package. this will call pacman on any AUR
based packages and remove them (if installed).
.TP
\fB\-Q\fR, \fB\-\-query\fR
query package database. this option is used to find
out what AUR packages are currently installed on the
system.
.TP
\fB\-u\fR, \fB\-\-upgrades\fR
perform an upgrade of installed packages on the the
system. this will attempt to upgrade all AUR installed
packages. a list of target packages may also be
passed.
.TP
\fB\-s\fR, \fB\-\-search\fR
search for packages in the AUR. the AUR rpc endpoints
will be called to attempt to find a package with a
name or description matching the input string (change
fields using \fB\-\-rpc\-field\fR).
.TP
\fB\-c\fR, \fB\-\-clean\fR
clean the cache. this will clean the naaman cache area
of any cache files. this can be used to
invalidate/remove old cache information for deprecated
packages or to reset duration caching options.
.TP
\fB\-d\fR, \fB\-\-deps\fR
naaman will attempt to build a dependency chain for
each package specified. upon completion naaman will
attempt to install (after confirmation) the determined
dependencies for the package.
.TP
\fB\-\-version\fR
display version information about naaman
.TP
\fB\-\-no\-sudo\fR
disable calling sudo. by default when naaman has to
call pacman directly (e.g. \fB\-R\fR), it will call with sudo
(if required). passing this option will prevent naaman
from using sudo.
.TP
\fB\-\-verbose\fR
verbose output. this setting will change the output
formatting and enable DEBUG level output. use this to
begin to debug naaman and to see more detail during
operations.
.TP
\fB\-\-trace\fR
trace debug logging. this option is useful to dump
extensive naaman logging information for indepth
troubleshooting or debugging purposes.
.TP
\fB\-\-pacman\fR PACMAN
pacman config. when naaman passes a configuration file
to pacman for initialization (via pyalpm) this
requires a pacman configuration file. this is NOT
passed when calling pacman directly (e.g. \fB\-R\fR).
.TP
\fB\-\-config\fR CONFIG
naaman config. specify the (optional) naaman
configuration file to use. see naaman.conf for
available options. naaman will read configs in the
order of (all optional): \fI\,/etc\/\fP, XDG_CONFIG_HOME, and
then \fB\-\-config\fR
.TP
\fB\-\-no\-confirm\fR
naaman will not ask for confirmation. when performing
install, update, and remove operations naaman will ask
for the user to confirm the operation. to disable
these prompts pass this option. this option will not
suppress makepkg or pacman prompts (use REMOVAL or
MAKEPKG in the naaman.conf)
.TP
\fB\-q\fR, \fB\-\-quiet\fR
quiet various parts of naaman to display less. certain
operations will display more AUR package information
that will be quiet (displaying minimal package
information, e.g. just package name) if this option is
provided.
.TP
\fB\-\-cache\-dir\fR CACHE_DIR
cache dir for naaman. naaman stores caching and
logging information to this location. naaman will
(attempt) to create this directory if it does not
exist.
.TP
\fB\-\-no\-config\fR
do not load the config file. to prevent naaman from
loading the configuration file when running specify
this option. this can allow running a specific
instance of naaman without certain config options
being loaded.
.TP
\fB\-\-builds\fR BUILDS
the location where naaman will perform builds. if not
set this will be in XDG_CACHE_HOME. specifying this
option will move where makepkg operations are
performed on the system.
.SS "Sync/Update options:"
.TP
\fB\-\-ignore\-for\fR N [N ...]
ignore packages for periods of time (hours).
specifying this options allows for ignoring certain
packages over a known duration. this is specified as
<package>=<hours>, where <package> will only check for
updates every <hours> period.
.TP
\fB\-\-vcs\-ignore\fR VCS_IGNORE
time betweeen vcs update checks (hours). specifying
this option will result in vcs\-based AUR packages to
only be updated every <hour> threshold. default is 720
(30 days)
.TP
\fB\-\-no\-vcs\fR
perform all sync operations but skip updating any vcs
packages. this will allow for performing various sync
operations without always having vcs packages
attempting to update.
.TP
\fB\-y\fR, \fB\-\-refresh\fR
refresh non\-vcs packages if there are updates in the
AUR for the package. packages with detected updates in
the AUR will be refreshed.
.TP
\fB\-yy\fR, \fB\-\-force\-refresh\fR
similar to \fB\-y\fR but will force refresh over any
\fB\-\-ignore\fR, \fB\-\-ignore\-for\fR, \fB\-\-vcs\-ignore\fR and disable any
rpc caching. Use this option to override the mentioned
flags and force updating or installing a package set.
.TP
\fB\-\-ignore\fR N [N ...]
ignore packages by name. packages ignored will be
skipped during upgrades unless forced (\fB\-yy\fR). utilize
this to keep packages back and/or prevent upgrading
packages
.TP
\fB\-\-no\-cache\fR
skip caching package files. by default naaman will
take the resulting makepkg output packages and place
them in the pacman cache directory. setting this will
disable this operation.
.TP
\fB\-\-skip\-deps\fR
skip dependency checks. naaman will attempt to detect
and error when checking an AUR package for other AUR
packages that are NOT installed. setting this will
disable this check.
.TP
\fB\-\-reorder\-deps\fR
by default naaman will reorder the input package set
install order if it detects certain packages must be
installed others given as input (e.g. \fB\-S\fR package1
package0 where package1 relies on package0 will change
to \fB\-S\fR package0 package1). to disable this behavior,
set this flag.
.TP
\fB\-\-rpc\-cache\fR RPC_CACHE
enable rpc caching (minutes). instead of making a web
rpc request to check for updated AUR package
information, naaman will cache the last received
information about a package for this duration of time.
default is 60 minutes.
.TP
\fB\-i\fR, \fB\-\-info\fR
display additional information about packages when
searching for information in the AUR. this can only be
used during a search but will present as much
information as possible to the user about the result
package set. passing multiple i parameters will
increase verbosity (e.g. \fB\-ii\fR)
.TP
\fB\-\-vcs\-install\-only\fR
by default when attempting a force update (\fB\-yy\fR) naaman
will attempt to download/clone the package and
determine the version for vcs packages. by settings
this flag when force updating naaman will skip the vcs
version check and force\-install the current version of
the vcs package regardless of what is currently
installed. disabling this would save bandwidth
.TP
\fB\-yyy\fR, \fB\-\-force\-force\-refresh\fR
similar to \fB\-yy\fR but will force force refresh over any
speciality checking (e.g. \fB\-\-vcs\-install\-only\fR). Use
this flag to update all AUR packages on the system
.TP
\fB\-f\fR, \fB\-\-fetch\fR
normally naaman will manage source retrieval and
package building without problem. though this relies
on knowing a PKGBUILD is safe to use and install.
utilize this option to tell naaman to fetch to a
location and not manage the files/perform the build.
the user can then verify the PKGBUILD and run makepkg
manually.
.TP
\fB\-\-fetch\-dir\fR FETCH_DIR
specifies the location to store fetched (\fB\-\-fetch\fR)
packages on the file system. whenever fetch is used
this is the directory that naaman will write to
(defaults to '.').
.TP
\fB\-\-rpc\-field\fR {name\-desc,name,maintainer}
when querying the AUR RPC endpoint, naaman will use a
default search field to search for packages. by
setting this argument naaman will be instructed to,
instead, search using the specified field when
querying the RPC endpoint during a search.
.TP
\fB\-\-do\-not\-track\fR N [N ...]
specify package names (1 or more) that naaman is NOT
responsible for tracking and should skip during
processing. naaman will not attempt to process these
packages at any point.
.TP
\fB\-\-makedeps\fR
include the make dependencies as part of the
dependency resolution when handling/resolving
dependencies.
.SS "Query options:"
.TP
\fB\-g\fR, \fB\-\-gone\fR
specifying this option will check for packages
installed from the AUR but are no longer in the AUR
(orphans).
.SH "SEE ALSO"
.B man naaman.conf


naaman/naaman.py
#!/usr/bin/python
"""
N(ot) A(nother) A(UR) Man(ager).

Is an AUR wrapper/manager that uses pacman as it's backing data store.
"""
import argparse
import os
import json
import shutil
import naaman.arguments.common as common_args
import naaman.arguments.config as config_args
import naaman.arguments.custom as csm_args
import naaman.arguments.utils as util_args
import naaman.arguments.query as query_args
import naaman.arguments.syncup as sync_args
import naaman.shell as sh
import naaman.aur as aur
import naaman.context as nctx
import naaman.logger as log
import naaman.consts as cst
from datetime import datetime, timedelta


def _validate_options(args, unknown, groups):
    """Validate argument options."""
    valid_count = 0
    invalid = False
    need_targets = False
    call_on = None

    if (args.search or args.query) and not args.verbose:
        def no_call(name):
            pass
        call_on = no_call
    else:
        def action(name):
            log.console_output("performing {}".format(name))
        call_on = action

    if args.sync:
        call_on("sync")
        valid_count += 1
        if args.upgrades or args.search or args.clean or \
           args.deps or args.fetch:
            sub_count = 0
            sub_command = None
            if args.upgrades:
                sub_command = "upgrade"
                sub_count += 1
            if args.clean:
                sub_command = "clean"
                sub_count += 1
            if args.search:
                sub_count += 1
            if args.deps:
                sub_count += 1
                sub_command = "deps"
            if args.fetch:
                sub_count += 1
                sub_command = "fetch"
            if sub_count == 1:
                if sub_command is not None:
                    call_on("sync: {}".format(sub_command))
            else:
                log.console_error("cannot perform multiple sub-options")
                invalid = True
        if args.search or args.deps or args.fetch or \
           (not args.upgrades and not args.clean and not args.deps):
            need_targets = True

    if args.remove:
        call_on("remove")
        valid_count += 1
        need_targets = True

    if args.query:
        call_on("query")
        valid_count += 1

    if not invalid:
        if valid_count > 1:
            log.console_error("multiple top-level arguments given")
        elif valid_count == 0:
            log.console_error("no valid top-level arguments given")
        if valid_count != 1:
            invalid = True

    if not invalid and \
       (args.search or args.upgrades or args.clean or args.fetch):
        if not args.sync:
            log.console_error(
                "search, upgrade, clean, and fetch are sync only")
            invalid = True

    if not invalid and args.info and not args.search:
        log.console_error("info only works with search")
        invalid = True

    if not invalid and args.info and args.quiet:
        log.console_error("info and quiet do not work together")
        invalid = True

    if not invalid and args.gone and not args.query:
        log.console_error("gone only works with query")
        invalid = True

    if not invalid and need_targets:
        if len(unknown) == 0:
            log.console_error("no targets specified")
            invalid = True

    if not args.pacman or not os.path.exists(args.pacman):
        log.console_error("invalid config file")
        invalid = True

    ctx = nctx.Context(unknown, groups, args)
    callback = None
    if not invalid:
        if args.query:
            if args.gone:
                callback = _gone
            else:
                callback = _query
        if args.search:
            callback = _search
        if args.upgrades:
            callback = _upgrades
        if args.clean:
            callback = _clean
        if args.deps:
            callback = _deps
        if args.sync:
            if not args.search and \
               not args.upgrades and \
               not args.clean and \
               not args.deps:
                callback = _sync
        if args.remove:
            callback = _remove

    if not invalid and callback is None:
        log.console_error("unable to find callback")
        invalid = True

    if invalid:
        ctx.exiting(1)
    callback(ctx)


def _resolution_output(context, name):
    """Output a dep resolution message."""
    if not context.quiet:
        log.update_progress("dependency resolution: {}".format(name))


def _load_deps(depth, packages, context, cache, parent):
    """Load dependencies for a package."""
    if packages is None or len(packages) == 0:
        return
    for p in packages:
        if p in cache:
            parent.add(cache[p])
            continue
        log.debug("resolving dependencies level {}, {}".format(depth, p))
        dependency = aur.deps_compare(p)
        p = dependency.pkg
        if context.check_pkgcache(p, dependency.version):
            continue
        pkg = _rpc_search(p, True, context, include_deps=True)
        if pkg is None:
            log.debug("non-aur {}".format(p))
            continue
        t = aur.DepTree(p)
        _resolution_output(context, p)
        _load_deps(depth + 1, pkg.deps, context, cache, t)
        parent.add(t)
        cache[p] = t


def _deps(context):
    """Handle dependency resolution."""
    log.debug("attempt dependency resolution")
    context.deps = False
    targets = context.targets
    for target in targets:
        resolved = aur.DepTree(target)
        log.debug("resolving {}".format(target))
        pkg = _rpc_search(target, True, context, include_deps=True)
        if pkg is None:
            log.console_error("unable to find package: {}".format(target))
            continue
        if pkg.deps is not None and len(pkg.deps) > 0:
            cache = {}
            visits = {}
            _load_deps(1, pkg.deps, context, cache, resolved)
            _resolution_output(context, "{} (complete)".format(target))
            actual = reversed(sorted(resolved.get(visits), key=lambda x: x[0]))
            context.targets = []
            for act in actual:
                log.debug(act)
                a = act[1]
                if a not in context.targets:
                    context.targets.append(a)
        else:
            log.debug("no deps...")
            context.targets = [target]
        log.trace(resolved)
        _sync(context)


def _clean(context):
    """Clean cache files."""
    log.debug("cleaning requested")
    files = [x for x in context.get_cache_files()]
    if len(files) == 0:
        log.console_output("no files to cleanup")
    else:
        _confirm(context, "clear cache files", [x[0] for x in files])
        for f in files:
            log.console_output("removing {}".format(f[0]))
            os.remove(f[1])
    dirs = [x for x in context.get_cache_dirs()]
    if len(dirs) == 0:
        log.console_output("no directories to cleanup")
    else:
        _confirm(context, "clear cache directories", [x for x in dirs])

        def remove_fail(func, path, err):
            log.debug("failed on removal")
            log.debug(path)
            log.debug(err)
            log.console_error("unable to cleanup {}".format(path))
        for d in dirs:
            shutil.rmtree(d, onerror=remove_fail)


def _confirm(ctx, message, package_names, default_yes=True):
    """Confirm package changes."""
    exiting = sh.confirm(message, package_names, default_yes, ctx.confirm)
    if exiting is not None:
        log.console_error("{}cancelled".format(exiting))
        ctx.exiting(1)


def _sync(context):
    """Sync packages (install)."""
    _syncing(context, True, context.targets, False)


def _get_deltahours(input_str, now):
    """Get a timedelta in hours."""
    log.debug("timedelta (hours)")
    last = datetime.fromtimestamp(float(input_str))
    seconds = (now - last).total_seconds()
    minutes = seconds / 60
    hours = minutes / 60
    log.trace((last, seconds, minutes, hours))
    return hours


def _check_vcs_ignore(context, threshold):
    """VCS caching check."""
    log.debug("checking vcs ignore cache")
    cache_check = context.cache_file("vcs")
    update_cache = True
    result = None
    now = context.now
    current_time = context.timestamp
    # we have a cache item, has necessary time elapsed?
    if os.path.exists(cache_check):
        with open(cache_check, 'r') as f:
            hours = _get_deltahours(f.read(), now)
            if hours < threshold:
                update_cache = False
                result = True
    if update_cache:
        log.debug("updating vcs last cache time")
        with open(cache_check, 'w') as f:
            f.write(str(current_time))
    return result


def _ignore_for(context, ignore_for, ignored):
    """Ignore for settings."""
    log.debug("checking ignored packages.")
    ignore_file = context.cache_file("ignoring")
    ignore_definition = {}
    now = context.now
    current_time = context.timestamp
    if os.path.exists(ignore_file):
        with open(ignore_file, 'r') as f:
            ignore_definition = json.loads(f.read())
    log.trace(ignore_definition)
    for i in ignore_for:
        if "=" not in i:
            log.warn("invalid ignore definition {}".format(i))
            continue
        parts = i.split("=")
        if len(parts) != 2:
            log.warn("invalid ignore format {}".format(i))
        package = parts[0]
        hours = 0
        try:
            hours = int(parts[1])
            if hours < 1:
                raise Exception("hour must be >= 1")
        except Exception as e:
            log.warn("invalid hour value {}".format(i))
            log.trace(e)
            continue
        update = True
        if package in ignore_definition:
            last = _get_deltahours(float(ignore_definition[package]), now)
            if last < hours:
                ignored.append(package)
                update = False
        if update:
            ignore_definition[package] = str(current_time)
    with open(ignore_file, 'w') as f:
        log.debug("writing ignore definitions")
        f.write(json.dumps(ignore_definition))


def _syncing(context, is_install, targets, updating):
    """Sync/install packages."""
    if context.root:
        log.console_error(
            "can not run install/upgrades as root (uses makepkg)")
        context.exiting(1)
    args = context.groups[sync_args.SYNC_UP_OPTIONS]
    ignored = args.ignore
    skip_filters = False
    if args.force_refresh or is_install:
        log.debug('skip filtering options')
        skip_filters = True
    if not ignored or skip_filters:
        ignored = []
    no_vcs = False
    if args.no_vcs or (args.refresh and not skip_filters):
        no_vcs = True
    if not no_vcs and args.vcs_ignore > 0 and not skip_filters:
        context.lock()
        try:
            if _check_vcs_ignore(context, args.vcs_ignore) is not None:
                log.trace("vcs ignore threshold met")
                no_vcs = True
        except Exception as e:
            log.error("unexpected vcs error")
            log.error(e)
        context.unlock()
    log.debug("novcs? {}".format(no_vcs))
    if args.ignore_for and len(args.ignore_for) > 0 and not skip_filters:
        log.debug("handling ignorefors")
        context.lock()
        try:
            _ignore_for(context, args.ignore_for, ignored)
        except Exception as e:
            log.error("unexpected ignore_for error")
            log.error(e)
        context.unlock()
    log.trace("ignoring {}".format(ignored))
    check_inst = []
    for name in targets:
        if name in ignored:
            log.console_output("{} is ignored".format(name))
            continue
        vcs = aur.is_vcs(name)
        if no_vcs and vcs:
            log.debug("skipping vcs package {}".format(name))
            continue
        package = _rpc_search(name, True, context)
        if package and package.name in context.do_not_track:
            log.debug("do not track: {}".format(package.name))
            continue
        if package:
            if vcs and \
               not args.vcs_install_only and \
               args.force_refresh and \
               not args.force_force_refresh:
                log.debug("checking vcs version")
                pkg = context.db.get_pkg(package.name)
                if pkg:
                    if not aur.check_vcs(package, context, pkg.version):
                        continue
                else:
                    log.debug("unable to find installed package...")
            check_inst.append(package)
        else:
            if updating:
                log.console_output(
                    "untracked package: {} (ignored)".format(name))
            else:
                log.console_error("unknown AUR package: {}".format(name))
                context.exiting(1)
    inst = []
    for item in context.reorders:
        obj = [x for x in check_inst if x.name == item]
        if len(obj) > 0:
            inst += obj
    for item in check_inst:
        obj = [x for x in inst if x.name == item.name]
        if len(obj) == 0:
            inst += [item]
    log.trace(inst)
    report = []
    do_install = []
    for i in inst:
        pkg = context.db.get_pkg(i.name)
        vers = i.version
        tag = ""
        vcs = aur.is_vcs(i.name)
        if pkg:
            if pkg.version == i.version or vcs:
                if not vcs and updating:
                    continue
                tag = " [installed]"
        else:
            if not is_install:
                log.console_error("{} not installed".format(i.name))
                context.exiting(1)
        log.trace(i)
        if vcs:
            vers = vcs
        report.append("{} {}{}".format(i.name, vers, tag))
        do_install.append(i)
    if len(do_install) == 0:
        log.console_output("nothing to do")
        context.exiting(0)
    _confirm(context, "install packages", report)
    makepkg = context.get_custom_arg(csm_args.CUSTOM_MAKEPKG)
    log.debug("makepkg {}".format(makepkg))
    cache = context.handle.cachedirs
    cache_dirs = ""
    if not args.no_cache and cache and len(cache) > 0:
        use_caches = []
        for c in cache:
            if " " in c:
                log.warn("cache dir with space is skipped ({})".format(c))
                continue
            use_caches.append(c)
        naaman_pkg = context.get_cache_pkgs()
        if not os.path.exists(naaman_pkg):
            os.makedirs(naaman_pkg)
        use_caches.append(naaman_pkg)
        cache_dirs = " ".join(['{}'.format(x) for x in use_caches])
    context.lock()
    try:
        for i in do_install:
            if not aur.install(i,
                               makepkg,
                               cache_dirs,
                               context,
                               None):
                log.console_error(
                    "error installing package: {}".format(i.name))
                next_pkgs = []
                after = False
                for e in do_install:
                    if e.name == i.name:
                        after = True
                        continue
                    if not after:
                        continue
                    next_pkgs.append(e.name)
                if len(next_pkgs) > 0:
                    _confirm(context,
                             "attempt to continue",
                             next_pkgs,
                             default_yes=False)
    except Exception as e:
        log.error("unexpected install error")
        log.error(e)
    context.unlock()


def _upgrades(context):
    """Ordered upgrade."""
    pkgs = list(_do_query(context))
    deps = aur.get_deps(pkgs)
    names = []
    for d in deps:
        if d.name in names:
            continue
        names.append(d.name)
    _syncing(context, False, names, True)


def _remove(context):
    """Remove package."""
    p = list(_do_query(context))
    _confirm(context,
             "remove packages",
             ["{} {}".format(x.name, x.version) for x in p])
    removals = context.get_custom_arg(csm_args.CUSTOM_REMOVAL)
    call_with = ['-R']
    if len(removals) > 0:
        call_with = call_with + removals
    call_with = call_with + [x.name for x in p]
    result = context.pacman(call_with)
    if not result:
        log.console_error("unable to remove packages")
        context.exiting(1)
    log.console_output("packages removed")


def _rpc_search(package_name, exact, context, include_deps=False):
    """Search for a package in the aur."""
    return aur.rpc_search(package_name, exact, context, include_deps)


def _search(context):
    """Perform a search."""
    if len(context.targets) != 1:
        log.console_error("please provide ONE target for search")
        context.exiting(1)
    for target in context.targets:
        log.debug("searching for {}".format(target))
        _rpc_search(target, False, context)


def _query(context):
    """Perform query."""
    _querying(context, False)


def _gone(context):
    """Perform query for dropped aur packages."""
    _querying(context, True)


def _querying(context, gone):
    """Querying for package information."""
    matched = False
    for q in _do_query(context):
        found = _rpc_search(q.name, True, context)
        if found:
            if gone:
                continue
        else:
            if not gone:
                continue
        if context.quiet:
            output = "{}"
        else:
            output = "{} {}"
        log.info(output.format(q.name, q.version))
        matched = True
    if not matched and not context.quiet:
        log.console_output("no packages found")


def _is_aur_pkg(pkg, sync_packages):
    """Detect AUR package."""
    return pkg.name not in sync_packages


def _do_query(context):
    """Query pacman."""
    syncpkgs = context.get_packages()
    if len(context.targets) > 0:
        for target in context.targets:
            pkg = context.db.get_pkg(target)
            valid = False
            if pkg and _is_aur_pkg(pkg, syncpkgs):
                    yield pkg
            else:
                log.console_error("unknown package: {}".format(target))
                context.exiting(1)
    else:
        for pkg in context.db.pkgcache:
            if _is_aur_pkg(pkg, syncpkgs):
                yield pkg


def main():
    """Entry point."""
    cache_dir = config_args.get_default_cache()
    config_file = config_args.get_default_config()
    parser = common_args.build(config_file, cache_dir)
    sync_args.sync_up_options(parser)
    query_args.options(parser)
    args, unknown = parser.parse_known_args()
    log.init(args.verbose, args.trace, args.cache_dir)
    log.trace("files/folders")
    log.trace(args.cache_dir)
    log.trace(args.config)
    if args.no_config:
        log.debug("not loading config")
    else:
        loaded = []
        for f in ["/etc/" + cst.CONFIG, config_file, args.config]:
            if f in loaded:
                continue
            loaded.append(f)
            args = config_args.load_config(args, f)
    util_args.manual_args(args)
    arg_groups = {}
    dirs = dir(args)
    custom_args = {}
    for k in csm_args.DEFAULT_OPTS:
        if k not in dirs:
            log.debug('setting default for {}'.format(k))
            setattr(args, k, csm_args.DEFAULT_OPTS[k])
        custom_args[k] = getattr(args, k)
    arg_groups[csm_args.CUSTOM_ARGS] = custom_args
    for group in parser._action_groups:
        g = {a.dest: getattr(args, a.dest, None) for a in group._group_actions}
        arg_groups[group.title] = argparse.Namespace(**g)
    log.trace(arg_groups)
    _validate_options(args, unknown, arg_groups)


if __name__ == "__main__":
    main()
